### DISCUSSION TITLE
jsPsych code runs on local server but not when inside JATOS

### DISCUSSION THREAD

**Discussion by amunn** [jspsych/jsPsych#2568] [Q&A]
I have a jsPsych experiment which has some browser checking code I wrote. This code works find running the jsPsych code locally or on cognition.run, but when I try to run the same code through a JATOS server it fails.

In the code I define an isBrowser function. (full code is [here](https://github.com/jspsych/jsPsych/discussions/1984))

```
function isBrowser(name) {
    return navigator.userAgent.toLowerCase().indexOf(name) != -1;
}
```
When I run the experiment through JATOS I get the following error:

```Uncaught (in promise) TypeError: isBrowser is not a function```

Is there something about wrapping the jsPsych code in the jatos wrapper that causes this? If so, how do I fix it?

(As a workaround, I can use the `browser-check` plugin, but I'd like to understand the problem more generally, since it seems to be related to how functions work, not this particular function.)

**Comment by jodeleeuw** [comment]
Hi @amunn,

This sounds like a scoping issue but I think we'd need to see the context of where the `isBrowser` function is defined and how the script(s) are loaded when running through JATOS. I don't think the code you've linked to shows where the function is defined relative to the context.

**Reply by becky-gilbert** [reply]
Hi @amunn, just to add to Josh's response about this being a scoping issue: I'm guessing the reason you might be having scoping problems that only occur with JATOS is because of the addition of the `jatos.onLoad` function. This function can change the scope and availability of different variables, depending on where you put it. If you haven't done so already, I would try putting _all_ of your code inside the `jatos.onLoad` function and see if that works. 

If you'd like more guidance/explanation, feel free to post more code as Josh suggested so that we can see where the function is created and called. You may also be interested in reading posts about JavaScript scope, such as [this one](https://dev.to/sandy8111112004/javascript-introduction-to-scope-function-scope-block-scope-d11) and [this one](https://dmitripavlutin.com/javascript-scope/).
